_posts/2018-03-19-making-a-components-pod.markdown

```diff
@@ -0,0 +1,703 @@
+}
+
+@end
+```
+
+As the `pod lib create` template uses storyboards, you will need to open up the example app's storyboard and [change
+the initial view controller](/images/making_cp_pod/settings_welcome.png) to be a `GDWelcomeViewController`. If you see a [white screen](/images/making_cp_pod/not_working.png) on the app launches then
+this hasn't been done. Run the app in the simulator, and you should get this screen:
+
+<center><img src="/images/making_cp_pod/success.png" width="50%" /></center>
+
+This is the default screen from the React Native template, and it's proof that everything has worked for our dev app.
+
+Let's take a second to re-cover what has happened to get to this point.
+
+1.  We used the `pod lib create` template to make a library repo
+
+2.  We used `react-native init` to make a React Native environment, which has the settings in the root and the
+    source code inside `src`
+
+3.  We've bundled the React Native code into our CocoaPod's asset folder
+
+4.  We set up the Podspec for GitDawg, and then the Podfile for the example project to consume it
+
+5.  We added [cocoapods-fix-react-native][cpfrn] to hot-fix the native files
+
+6.  We added a UIViewController for the default screen from `react-native init` to our CocoaPod, and ran
+    `bundle exec pod install` to update the example project
+
+7.  We changed the storyboard reference to point to the UIViewController from our Pod, and ran the simulator to see
+    our welcome screen
+
+This is a full run-through of how your Pod would look when integrated into your main app's codebase. At this point
+you have a unique, isolated app which is going to be your development environment. In our case this app is a menu of
+different root screens and admin flags.
+
+—
+
+OK, let’s go take this and migrate it into GitHawk. This is our end-goal:
+
+<center><img src="/images/making_cp_pod/githawk.gif" width="75%" /></center>
+
+Our setup is going to be different here because we can't rely on React Native coming from the file-system, as we
+want to make sure our app has no hint of JS tooling. So we will use CocoaPods to handle downloading and setting up
+our versions of the React Native libraries. As of 0.54.x, that is React and Yoga.
+
+We want to have a local copy of the JSON version of Podspecs for each of these. They can be generated from the
+Podspecs using `bundle exec pod ipc spec [file.podspec]`. Let's generate one for React:
+
+```sh
+$ cd GitDawg/node_modules/react-native/; pod ipc spec React.podspec
+```
+
+It will output a bunch of JSON to your terminal. This is perfect. Let's move that text to a file on your desktop.
+
+```sh
+$ pod ipc spec node_modules/react-native/React.podspec > ~/Desktop/React.podspec.json
+```
+
+You'll see no output if everything went fine. Before you grab that podspec, let's get the one for yoga too.
+
+```sh
+$ cd ReactCommon/yoga/; pod ipc spec yoga.podspec > ~/Desktop/yoga.podspec.json
```

Comment:
It looks like the podspec generated here is incorrect.

<details>
<summary>Generated Yoga podspec</summary>

```json
{
  "name": "yoga",
  "version": "0.54.4.React",
  "license": {
    "type": "BSD"
  },
  "homepage": "https://facebook.github.io/yoga/",
  "documentation_url": "https://facebook.github.io/yoga/docs/api/c/",
  "summary": "Yoga is a cross-platform layout engine which implements Flexbox.",
  "description": "Yoga is a cross-platform layout engine enabling maximum collaboration within your team by implementing an API many designers are familiar with, and opening it up to developers across different platforms.",
  "authors": "Facebook",
  "source": {
    "git": "https://github.com/facebook/react-native.git",
    "tag": "v0.54.4"
  },
  "module_name": "yoga",
  "requires_arc": false,
  "compiler_flags": [
    "-fno-omit-frame-pointer",
    "-fexceptions",
    "-Wall",
    "-Werror",
    "-std=c++1y",
    "-fPIC"
  ],
  "platforms": {
    "ios": "8.0",
    "tvos": "9.2"
  },
  "source_files": "yoga/**/*.{cpp,h}"
}
```

</details>

The `source_files` don't match any files in the repo at that tag, it looks like it's missing the `ReactCommon/` from the beginning of it. I tried adding it but then I ran into this problem: https://github.com/facebook/react-native/issues/17893 Happy to pair on this.

Reply:
I know the answer to that, because I made the PR 👍 - https://github.com/facebook/react-native/pull/18492/files

We need to make sure that when the yoga podspec is generated it includes the env var `"INSTALL_YOGA_WITHOUT_PATH_OPTION"`

Reply:
Yup, confirmed that corrected the generated podspec. Any idea about the missing `<algorithm>` `#include`?

Reply:
Hrm, I remember digging into this, but I don't remember having an answer - I feel like this fixes that `spec.public_header_files = 'yoga/{Yoga,YGEnums,YGMacros}.h'`

Reply:
Yeah, it looks like it's just that your PR isn't reflected in the `yoga.podspec` file in GitHawk's `nodule_modules`. I'll take a look.

Reply:
👍

Reply:
Okay, this worked after manually applying the changes from https://github.com/facebook/react-native/pull/18492 to the Ruby podspec (and including the env var with `pod ipc spec`). We should think about using 0.55 or whatever version that PR is included with. 

Ran into some more problems, though, with missing `RCTPackagerClient.h`, then missing `RCTValueAnimatedNode.h`. Going to take a break and come back to it later.